src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+
+        buf1[..U32_SIZE].copy_from_slice(&stream_id.to_le_bytes());
+        buf1[U32_SIZE..U32_SIZE + buffer_size].copy_from_slice(&frames[..buffer_size]);
+
+        buf2[..U32_SIZE].copy_from_slice(&stream_id.to_le_bytes());
+        buf2[U32_SIZE..U32_SIZE + buffer_size]
+            .copy_from_slice(&frames[buffer_size..buffer_size * 2]);
+
+        let mut outputs = vec![0; 32];
+        let mut token1 = unsafe { self.tx_queue.add(&[&buf1], &mut [&mut outputs]).unwrap() };
+        let mut token2 = unsafe { self.tx_queue.add(&[&buf2], &mut [&mut outputs]).unwrap() };
+
+        let xfer_times = if frames.len() % buffer_size == 0 {
+            frames.len() / buffer_size
+        } else {
+            frames.len() / buffer_size + 1
+        };
+
+        let mut turn1 = true;
+
+        for i in 2..xfer_times {
+            if self.tx_queue.should_notify() {
+                self.transport.notify(TX_QUEUE_IDX);
+            }
+
+            let start_byte = i * buffer_size;
+            let end_byte = if i != xfer_times - 1 {
+                (i + 1) * buffer_size
+            } else {
+                frames.len()
+            };
+            if turn1 {
+                while let Err(_) = unsafe {
+                    self.tx_queue
+                        .pop_used(token1, &[&buf1], &mut [&mut outputs])
+                } {
+                    spin_loop();
+                }
+                turn1 = false;
+                buf1[U32_SIZE..U32_SIZE + end_byte - start_byte]
+                    .copy_from_slice(&frames[start_byte..end_byte]);
+                token1 = unsafe { self.tx_queue.add(&[&buf1], &mut [&mut outputs]).unwrap() }
+            } else {
+                while let Err(_) = unsafe {
+                    self.tx_queue
+                        .pop_used(token2, &[&buf2], &mut [&mut outputs])
+                } {
+                    spin_loop();
+                }
+                turn1 = true;
+                buf2[U32_SIZE..U32_SIZE + end_byte - start_byte]
+                    .copy_from_slice(&frames[start_byte..end_byte]);
+                token2 = unsafe { self.tx_queue.add(&[&buf2], &mut [&mut outputs]).unwrap() }
+            }
+        }
+        // wait for the last buffer
+        if turn1 {
+            while let Err(_) = unsafe {
+                self.tx_queue
+                    .pop_used(token1, &[&buf1], &mut [&mut outputs])
+            } {
+                spin_loop();
+            }
+        } else {
+            while let Err(_) = unsafe {
+                self.tx_queue
+                    .pop_used(token2, &[&buf2], &mut [&mut outputs])
+            } {
+                spin_loop();
+            }
+        }
+        Ok(())
+    }
+
+    /// Transfer PCM frame to device, based on the stream type(OUTPUT/INPUT).
+    /// 
+    /// This is a non-blocking method that returns a token.
+    pub fn pcm_xfer_nb(&mut self, stream_id: u32, frames: &[u8]) -> Result<u16> {
```

Comment:
Calling this twice in quick succession will cause undefined behaviour as the same `self.output_rsp` is used for two entries in the virt queue, and the device may be writing to the second while we are reading from the first.

Reply:
Fixed.